kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/BigQueryStorageApiBatchSinkTaskTest.java

```diff
@@ -0,0 +1,171 @@
+
+    Storage storage = mock(Storage.class);
+    SinkTaskContext sinkTaskContext = mock(SinkTaskContext.class);
+    SchemaRetriever schemaRetriever = mock(SchemaRetriever.class);
+    SchemaManager schemaManager = mock(SchemaManager.class);
+    Map<TableId, Table> cache = new HashMap<>();
+    BigQuerySinkTask testTask = new BigQuerySinkTask(
+            bigQuery, schemaRetriever, storage, schemaManager, cache, mockedStorageWriteApiBatchStream, mockedBatchHandler);
+
+    final String topic = "test-topic";
+
+    Map<TopicPartition, OffsetAndMetadata> mockedOffset = mock(Map.class);
+
+    @Before
+    public void setUp()  {
+        reset(mockedStorageWriteApiBatchStream);
+        properties = propertiesFactory.getProperties();
+
+        properties.put(BigQuerySinkConfig.TOPICS_CONFIG, topic);
+        properties.put(BigQuerySinkConfig.USE_STORAGE_WRITE_API_CONFIG, "true");
+        properties.put(BigQuerySinkConfig.ENABLE_BATCH_MODE_CONFIG, "true");
+        properties.put(BigQuerySinkConfig.DEFAULT_DATASET_CONFIG, "scratch");
+        spoofedRecordOffset.set(0);
+        mockedOffset.put(new TopicPartition(topic, 0), new OffsetAndMetadata(0));
+
+        doNothing().when(mockedStorageWriteApiBatchStream).appendRows(any(), any(), eq("dummyStream"));
+        doNothing().when(mockedStorageWriteApiBatchStream).shutdown();
+        doNothing().when(mockedBatchHandler).createNewStream();
+        when(mockedBatchHandler.updateOffsetsOnStream(any(), any())).thenReturn("dummyStream");
+        when(mockedBatchHandler.getCommitableOffsets()).thenReturn(mockedOffset);
+        testTask.initialize(sinkTaskContext);
+        testTask.start(properties);
+    }
+
+    @After
+    public void teardown() {
+        testTask.stop();
+    }
+
+    @Test
+    public void testPut() {
+        testTask.put(Collections.singletonList(spoofSinkRecord()));
+        testTask.flush(Collections.emptyMap());
+
+        verify(mockedStorageWriteApiBatchStream, times(1)).appendRows(any(), any(), any());
+    }
+
+    @Test
+    public void testStart() throws InterruptedException {
+        Thread.sleep(12000); // 10 seconds is default, check after 12 seconds
+        verify(mockedBatchHandler, times(1)).createNewStream();
+    }
+
+    @Test(expected = BigQueryConnectException.class)
+    public void testSimplePutException() throws Exception {
+        BigQueryStorageWriteApiConnectException exception = new BigQueryStorageWriteApiConnectException("error 12345");
+
+        doThrow(exception).when(mockedStorageWriteApiBatchStream).appendRows(any(), any(), eq("dummyStream"));
+
+        testTask.put(Collections.singletonList(spoofSinkRecord()));
+        try {
+            while (true) {
+                Thread.sleep(100);
+                testTask.put(Collections.emptyList());
+            }
+        } catch (Exception e) {
+            assertTrue(e.getCause() instanceof BigQueryStorageWriteApiConnectException);
+            throw e;
```

Comment:
I am not sure when this code will be triggered. Is it not the case that it will throw error on previous line and this while loop won't get executed - `testTask.put(Collections.singletonList(spoofSinkRecord()));`?

Reply:
No actually. We execute threads in put method and line:127 here would raise an exception in one of the threads. We rely on this method ([maybeThrowEncounteredError](https://github.com/confluentinc/kafka-connect-bigquery/blob/691dddc5a1f9224df40588b9fbe239cd88b63775/kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java#L301) ) to see if any exception occurred in previous calls to put and fail task. Since the thread execution is async we cannot confidently say in which call to put will there be exception so the while loop. 
line:129 would not throw any exception.